File: extensions/codex_stream.py
"""Parse Codex exec --json output into structured harness events."""

from __future__ import annotations

import json
from typing import Any, Generator, Iterator

from claude_stream import _emit_image_events


def parse_codex_stream(lines: Iterator[str]) -> Generator[dict[str, Any], None, None]:
    parser = new_codex_stream_parser()
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            envelope = json.loads(line)
        except json.JSONDecodeError:
            continue
        yield from parser.handle(envelope)


def new_codex_stream_parser() -> "_CodexStreamParser":
    return _CodexStreamParser()


class _CodexStreamParser:
    def __init__(self) -> None:
        self.seen_tool_starts: set[str] = set()
        self.seen_tool_ends: set[str] = set()
        self.seen_tool_results: set[str] = set()

    def handle(self, envelope: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        t = envelope.get("type") or ""

        if t == "thread.started":
            thread_id = envelope.get("thread_id") or ""
            if thread_id:
                yield {"type": "session_id", "sessionId": thread_id}
            return

        if t in ("item.started", "item.updated", "item.completed"):
            yield from self._handle_item(t, envelope.get("item") or {})
            return

        if t == "error":
            msg = envelope.get("message") or (envelope.get("error") or {}).get("message") or ""
            if msg:
                yield {"type": "error", "error": msg}
            return

        if t == "turn.failed":
            msg = envelope.get("message") or (envelope.get("error") or {}).get("message") or "turn failed"
            yield {"type": "error", "error": msg}

    def _handle_item(self, event_type: str, item: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        item_type = item.get("type") or ""

        if item_type == "agent_message":
            if event_type == "item.completed":
                text = item.get("text") or ""
                if text:
                    yield {"type": "text", "content": text}
            return

        if item_type != "mcp_tool_call":
            return

        tool_id = item.get("id") or ""
        if not tool_id:
            return

        server = item.get("server") or ""
        tool = item.get("tool") or ""
        tool_name = f"{server}/{tool}" if server else tool

        if event_type == "item.started":
            if tool_id not in self.seen_tool_starts:
                self.seen_tool_starts.add(tool_id)
                yield {"type": "tool_call_start", "toolCallId": tool_id, "toolName": tool_name}
            return

        if event_type != "item.completed":
            return

        if tool_id not in self.seen_tool_starts:
            self.seen_tool_starts.add(tool_id)
            yield {"type": "tool_call_start", "toolCallId": tool_id, "toolName": tool_name}

        args = item.get("arguments")
        if args is None:
            args_json = "{}"
        elif isinstance(args, str):
            args_json = args
        else:
            args_json = json.dumps(args)
        yield {"type": "tool_call_args", "toolCallId": tool_id, "toolArgs": args_json}

        if tool_id not in self.seen_tool_ends:
            self.seen_tool_ends.add(tool_id)
            yield {
                "type": "tool_call_end",
                "toolCallId": tool_id,
                "toolName": tool_name,
                "toolArgs": args_json,
            }

        if tool_id in self.seen_tool_results:
            return
        self.seen_tool_results.add(tool_id)

        status = item.get("status") or ""
        err = (item.get("error") or {}).get("message") or ""
        if status == "failed" or err:
            yield {"type": "tool_call_result", "toolCallId": tool_id, "content": err or "tool failed"}
            return

        result = item.get("result")
        if result is not None:
            content = result if isinstance(result, str) else json.dumps(result)
            yield {"type": "tool_call_result", "toolCallId": tool_id, "content": content}
            yield from _emit_image_events(result)

File: internal/agent/codex_stream.go
// Copyright (c) Mehmet Bektas <mbektasgh@outlook.com>

package agent

import (
	"encoding/json"
	"fmt"
)

type codexStreamParser struct {
	seenToolStarts   map[string]struct{}
	seenToolEnds     map[string]struct{}
	seenToolResults  map[string]struct{}
}

func newCodexStreamParser() *codexStreamParser {
	return &codexStreamParser{
		seenToolStarts:  make(map[string]struct{}),
		seenToolEnds:    make(map[string]struct{}),
		seenToolResults: make(map[string]struct{}),
	}
}

func (p *codexStreamParser) handleLine(line []byte, events chan<- Event) (sessionID string, done bool) {
	if len(line) == 0 {
		return "", false
	}

	var envelope struct {
		Type     string `json:"type"`
		ThreadID string `json:"thread_id"`
		Message  string `json:"message"`
		Error    struct {
			Message string `json:"message"`
		} `json:"error"`
		Item json.RawMessage `json:"item"`
	}
	if err := json.Unmarshal(line, &envelope); err != nil {
		fmt.Printf("[codex stdout] %s\n", line)
		return "", false
	}

	switch envelope.Type {
	case "thread.started":
		return envelope.ThreadID, false
	case "item.started", "item.updated", "item.completed":
		p.handleItem(envelope.Type, envelope.Item, events)
	case "turn.completed":
		return "", true
	case "turn.failed":
		msg := envelope.Error.Message
		if msg == "" {
			msg = "turn failed"
		}
		events <- Event{Type: EventError, Error: msg}
	case "error":
		msg := envelope.Message
		if msg == "" {
			msg = envelope.Error.Message
		}
		if msg != "" {
			events <- Event{Type: EventError, Error: msg}
		}
	}
	return "", false
}

func (p *codexStreamParser) handleItem(eventType string, raw json.RawMessage, events chan<- Event) {
	if len(raw) == 0 || string(raw) == "null" {
		return
	}

	var item struct {
		ID        string          `json:"id"`
		Type      string          `json:"type"`
		Text      string          `json:"text"`
		Server    string          `json:"server"`
		Tool      string          `json:"tool"`
		Arguments json.RawMessage `json:"arguments"`
		Result    json.RawMessage `json:"result"`
		Error     struct {
			Message string `json:"message"`
		} `json:"error"`
		Status string `json:"status"`
	}
	if err := json.Unmarshal(raw, &item); err != nil {
		return
	}

	switch item.Type {
	case "agent_message":
		if eventType == "item.completed" && item.Text != "" {
			events <- Event{Type: EventText, Content: item.Text}
		}
	case "mcp_tool_call":
		p.handleMCPToolCall(eventType, item, events)
	}
}

func (p *codexStreamParser) handleMCPToolCall(eventType string, item struct {
	ID        string          `json:"id"`
	Type      string          `json:"type"`
	Text      string          `json:"text"`
	Server    string          `json:"server"`
	Tool      string          `json:"tool"`
	Arguments json.RawMessage `json:"arguments"`
	Result    json.RawMessage `json:"result"`
	Error     struct {
		Message string `json:"message"`
	} `json:"error"`
	Status string `json:"status"`
}, events chan<- Event) {
	toolID := item.ID
	if toolID == "" {
		return
	}

	toolName := item.Tool
	if item.Server != "" {
		toolName = item.Server + "/" + item.Tool
	}

	if eventType == "item.started" {
		if _, ok := p.seenToolStarts[toolID]; ok {
			return
		}
		p.seenToolStarts[toolID] = struct{}{}
		events <- Event{
			Type:       EventToolCallStart,
			ToolCallID: toolID,
			ToolName:   toolName,
		}
		return
	}

	if eventType != "item.completed" {
		return
	}

	if _, ok := p.seenToolStarts[toolID]; !ok {
		p.seenToolStarts[toolID] = struct{}{}
		events <- Event{
			Type:       EventToolCallStart,
			ToolCallID: toolID,
			ToolName:   toolName,
		}
	}

	argsJSON := string(item.Arguments)
	if argsJSON == "" || argsJSON == "null" {
		argsJSON = "{}"
	}
	events <- Event{
		Type:       EventToolCallArgs,
		ToolCallID: toolID,
		ToolArgs:   argsJSON,
	}

	if _, ok := p.seenToolEnds[toolID]; !ok {
		p.seenToolEnds[toolID] = struct{}{}
		events <- Event{
			Type:       EventToolCallEnd,
			ToolCallID: toolID,
			ToolName:   toolName,
			ToolArgs:   argsJSON,
		}
	}

	if _, ok := p.seenToolResults[toolID]; ok {
		return
	}
	p.seenToolResults[toolID] = struct{}{}

	if item.Status == "failed" || item.Error.Message != "" {
		msg := item.Error.Message
		if msg == "" {
			msg = "tool failed"
		}
		events <- Event{
			Type:       EventToolCallResult,
			ToolCallID: toolID,
			Content:    msg,
		}
		return
	}

	if len(item.Result) > 0 && string(item.Result) != "null" {
		events <- Event{
			Type:       EventToolCallResult,
			ToolCallID: toolID,
			Content:    string(item.Result),
		}
		emitImageEvents(item.Result, "", events)
	}
}
